### PR TITLE
Yield tuples by value when appropriate

### PR DESCRIPTION
**Pull request by vasslitvinov**
When yielding a tuple within an iterator with the default return intent, yield the components by value instead of by reference. This is to match the yield mode of yielding a component outside of a tuple, which is done by value. For example, this makes `yield (myRecord,1);` yield myRecord by value just like `yield myRecord;` already does.

Iterators that wish to opt out, i.e., use the prior semantics, can use FLAG_DONT_UNREF_FOR_YIELDS / pragma "do not unref for yields". We plan to replace this pragma with tuple per-component-refness specifiers, see #21717. Use this pragma in a couple of tests as in StencilDist where needed.

The standalone and follower iterators for loop expressions use the prior semantics. This is to allow the following loop body to update array elements, see `test/expressions/loop-expr/forall-over-zip-over-arrays.chpl` and https://github.com/chapel-lang/chapel/issues/10282#issuecomment-406243845 :

```chpl
forall xy in ([ab in zip(A,B)] ab) do
  xy = (0,0);
```

Implementation-wise, the compiler inserts a call to initCopy() on the tuple being yielded, which creates copies of the components.

While there, exit early from insertUnrefForArrayOrTupleReturn() when the return type is known to be neither an array nor a tuple, to avoid doing unnecessary work.
When the function is an iterator, process the yields then return.

The change in implementForallIntents.cpp converts an INT_ASSERT to a USR_FATAL.

Post-merge: add more testing.

### CODE REVIEW

File: compiler/resolution/resolveFunction.cpp [hunk 1]
```diff
@@ -1032,6 +1038,11 @@ static void insertUnrefForArrayOrTupleReturn(FnSymbol* fn) {
 
   Symbol* ret = fn->getReturnSymbol();
 
+  // surely returning neither tuple nor array, so return
+  if (ret->type != dtUnknown                   &&
+      ! ret->type->symbol->hasFlag(FLAG_ARRAY) &&
+      ! ret->type->symbol->hasFlag(FLAG_DOMAIN) ) return;
```
Comment (e-kayrakli) [inline compiler/resolution/resolveFunction.cpp:1044]: Should this be checking for `FLAG_TUPLE` also, or `dtUnknown` implies tuple when we run this code?

File: compiler/resolution/resolveFunction.cpp [hunk 2]
```diff
@@ -1158,6 +1169,51 @@ static CallExpr* findSetShape(CallExpr* setRet, Symbol* ret) {
   return NULL;
 }
 
+static void insertUnrefForTupleYields(FnSymbol* fn) {
+  // do not unref:
+  if (fn->retTag == RET_TYPE       ||  // could iterators yield types?
```
Comment (e-kayrakli) [inline compiler/resolution/resolveFunction.cpp:1174]: I certainly want to say "no" :)

**Comment by e-kayrakli** [comment]
> The standalone and follower iterators for loop expressions use the prior semantics. This is to allow the following loop body to update array elements, see test/expressions/loop-expr/forall-over-zip-over-arrays.chpl

OK with this for now, but do we want that snippet to work? Is the idea that you can iterate over a forall expression and that doesn't necessarily imply copying the data that it runs over? Even with that behavior, if I assign the expression to a `var` it'll be copied, right?

**Review by e-kayrakli — APPROVED** [review]
At just one location you may need a check for tuple-ness. It is good to go otherwise.

**Comment by vasslitvinov** [comment]
> > `forall xy in ([ab in zip(A,B)] ab) do xy = (0,0);`
> do we want that snippet to work?

That was my question, too. Turns out that the answer is "yes", see https://github.com/chapel-lang/chapel/issues/10282#issuecomment-406243845

To rephrase, `ab` is a pair of references in `forall ab in zip(A,B) do ...`. Therefore it is natural for them to remain references when we rewrite that loop as a loop expression `( forall ab in zip(A,B) do ab)` .

File: compiler/resolution/resolveFunction.cpp (hunk 1 above)
Comment (vasslitvinov) [reply compiler/resolution/resolveFunction.cpp:1045]: Indeed it should be checking FLAG_TUPLE. I added the check.

**Comment by bradcray** [comment]
Did CoMD and MiniMD not need to change after all?

**Comment by vasslitvinov** [comment]
> Did CoMD and MiniMD not need to change after all?

No, as it turns out. Only the `boundaries` iterators needed to be annotated with the "do not yield by value" pragma.

**Comment by vasslitvinov** [comment]
> > Only the boundaries iterators needed to be annotated
> 
>  those live in StencilDist.chpl?

Correct.

Now, there is a StencilDist in modules/dists/ and another one in the CoMD directory, called AccumStencilDist. I annotated the boundaries iterators in each. If there are other stencil dists, they did not need adjustments.

**Comment by bradcray** [comment]
Ah, I missed that CoMD was in studies... for some reason in our discussion the other day, I was imagining it to be in the release (so was surprised not to see any changes in the release dir).  Thanks!

**Pull request merged**